File: 62. ReverseNodesInKGroup.java
public class reverseNodesInKGroup {

    static class ListNode {
        int val;
        ListNode next;

        ListNode() {}

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public static ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k <= 1) {
            return head;
        }

        ListNode dummy = new ListNode(0, head);
        ListNode groupPrev = dummy;

        while (true) {
            ListNode kth = getKth(groupPrev, k);
            if (kth == null) {
                break;
            }

            ListNode groupNext = kth.next;
            ListNode prev = groupNext;
            ListNode curr = groupPrev.next;

            while (curr != groupNext) {
                ListNode temp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = temp;
            }

            ListNode oldGroupHead = groupPrev.next;
            groupPrev.next = kth;
            groupPrev = oldGroupHead;
        }

        return dummy.next;
    }

    private static ListNode getKth(ListNode start, int k) {
        while (start != null && k > 0) {
            start = start.next;
            k--;
        }
        return start;
    }

    private static ListNode build(int... values) {
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;
        for (int value : values) {
            tail.next = new ListNode(value);
            tail = tail.next;
        }
        return dummy.next;
    }

    private static void print(ListNode head) {
        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val);
            if (curr.next != null) {
                System.out.print(" -> ");
            }
            curr = curr.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode a = build(1, 2, 3, 4, 5);
        print(reverseKGroup(a, 2));

        ListNode b = build(1, 2, 3, 4, 5);
        print(reverseKGroup(b, 3));
    }
}

File: 62. ReverseNodesInKGroup.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode *newNode(int val) {
    struct ListNode *node = malloc(sizeof(*node));
    node->val = val;
    node->next = NULL;
    return node;
}

static struct ListNode *getKth(struct ListNode *start, int k) {
    while (start != NULL && k > 0) {
        start = start->next;
        k--;
    }
    return start;
}

struct ListNode *reverseKGroup(struct ListNode *head, int k) {
    if (head == NULL || k <= 1) return head;

    struct ListNode dummy;
    dummy.val = 0;
    dummy.next = head;
    struct ListNode *groupPrev = &dummy;

    while (1) {
        struct ListNode *kth = getKth(groupPrev, k);
        if (kth == NULL) break;

        struct ListNode *groupNext = kth->next;
        struct ListNode *prev = groupNext;
        struct ListNode *curr = groupPrev->next;

        while (curr != groupNext) {
            struct ListNode *temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        struct ListNode *oldGroupHead = groupPrev->next;
        groupPrev->next = kth;
        groupPrev = oldGroupHead;
    }

    return dummy.next;
}

static struct ListNode *build(int *values, int n) {
    struct ListNode dummy;
    dummy.next = NULL;
    struct ListNode *tail = &dummy;
    for (int i = 0; i < n; i++) {
        tail->next = newNode(values[i]);
        tail = tail->next;
    }
    return dummy.next;
}

static void printList(struct ListNode *head) {
    struct ListNode *curr = head;
    while (curr != NULL) {
        printf("%d", curr->val);
        if (curr->next != NULL) printf(" -> ");
        curr = curr->next;
    }
    printf("\n");
}

int main(void) {
    int v1[] = {1, 2, 3, 4, 5};
    struct ListNode *a = build(v1, 5);
    printList(reverseKGroup(a, 2));

    int v2[] = {1, 2, 3, 4, 5};
    struct ListNode *b = build(v2, 5);
    printList(reverseKGroup(b, 3));

    return 0;
}
